tests/test_math.hh

```diff
@@ -484,7 +484,7 @@ namespace rascal {
 
     ~Hyp1f1GradientProvider() = default;
 
-    Eigen::Ref<Eigen::Array<double, 1, Eigen::Dynamic>>
+    Eigen::Array<double, 1, Eigen::Dynamic>
```

Comment:
So `result_flat` is initialized as an `Eigen::Map<Eigen::Array>`, and then converted to a `Eigen::Array` when returning. Does this creates a copy of the data? If so, why not return a `MatrixXd` instead, or change the code to use `Eigen::Array` everywhere?

Reply:
Not as far as I can tell; the `Map` is just for reshaping the matrix (flattening, as the variable name would suggest) from the original shape of `n_max` by `l_max + 1` (rectangular) to the flat vector that the gradient test expects. The copy on return _should_ be optimized out via RVO into the corresponding vector in `test_gradients`: Either this
https://github.com/cosmo-epfl/librascal/blob/d9dd3e8721eddc8b014e95bcb2340d1e6134acb6/tests/test_math.hh#L340
or one of these
https://github.com/cosmo-epfl/librascal/blob/d9dd3e8721eddc8b014e95bcb2340d1e6134acb6/tests/test_math.hh#L376-L379

So there is _one_ copy that happens, from the Hyp1f1 result array into the local variable of `test_gradients`. But if you look at the gradient testing code you'll see why it's necessary to have that one copy: We need to call `f()` with different input values and compare the results (for a finite-difference analysis). But if `f()` just returns some sort of reference or map to the underlying data in `hyp1f1` then we get big problems because the underlying data changes with each call.

If your question was instead about `Matrix` versus `Array`, then no, there's no copying that happens when converting from one to the other; they're just different interfaces to the same data. Eigen will be able to tell the compiler to optimize out most of the "stupid copies"; better to have clearer expressions that better communicate our intent than horrific un-debuggable errors later on.